src/plugins/intel_cpu/src/nodes/mkldnn_non_zero.cpp

```diff
@@ -108,32 +115,148 @@ void MKLDNNNonZeroNode::executeSpecified() {
     auto dstMemPtr = getChildEdgeAt(0)->getMemoryPtr();
     Shape inShape = getParentEdgeAt(0)->getMemory().GetShape();
     size_t inRank = inShape.getRank();
-    size_t nonZeroCount = getNonZeroElementsCount(src, inShape);
+    std::vector<size_t> nonZeroCounts = getNonZeroElementsCount(src, inShape);
+    std::vector<size_t> destIndices(nonZeroCounts);
+    size_t totalNonZeroCount = 0;
+
+    for (size_t i = 0; i < nonZeroCounts.size(); ++i) {
+        destIndices[i] = totalNonZeroCount;
+        totalNonZeroCount += nonZeroCounts[i];
+    }
 
     if (isDynamicNode()) {
-        VectorDims newDims{inRank, nonZeroCount};
+        VectorDims newDims{inRank, totalNonZeroCount};
         redefineOutputMemory({newDims});
     }
     int *dst = reinterpret_cast<int *>(dstMemPtr->GetPtr());
-    size_t inSize = inShape.getElementsCount();
-    auto srcStrides = getParentEdgeAt(0)->getMemory().GetDescWithType<BlockedMemoryDesc>()->getStrides();
-    if (nonZeroCount == 0)
+    auto srcDims = inShape.getDims();
+    if (totalNonZeroCount == 0)
         return;
-    if (inShape.getRank() == 0) {
+
+    switch (inRank) {
+    case 0:
         dst[0] = 0;
-    } else {
-        size_t colIndex = 0, outIndex = 0;
-        for (size_t i = 0; i < inSize; i++) {
-            if (src[i] != zero) {
-                size_t temp = i;
-                for (size_t j = 0; j < inRank; j++) {
-                    outIndex = j * nonZeroCount + colIndex;
-                    dst[outIndex] = temp / srcStrides[j];
-                    temp = temp % srcStrides[j];
-                }
-                colIndex++;
-            }
-        }
+        break;
+    case 1:
+        parallel_for(srcDims[0],
+                     [&](size_t ithr, size_t i0) {
+                         size_t inputIndex = i0;
+                         size_t& outputIndex = destIndices[ithr];
+
+                         if (src[inputIndex] != zero) {
+                             dst[outputIndex] = static_cast<int>(i0);
```

Comment:
Declare srcDims, ithr, i0 as int to avoid type casting here.

Reply:
Partially done:
- for srcDims std::transform() was used, since the original vector uses size_t type for elements;
- ithr is hardcoded as size_t in src/inference/include/ie/ie_parallel.hpp on the line 372, so I kept it as is.